File: logic/src/main/java/com/tp/holdem/logic/model/HandRank.java
package com.tp.holdem.logic.model;

import com.tp.holdem.model.common.Hands;
import io.vavr.collection.List;
import lombok.Value;

@Value
public class HandRank {
	private Hands hand;
	private List<Card> cardsThatMakeHand;

	public static HandRank from(Hands hand, List<Card> cards) {
		return new HandRank(hand, cards);
	}

	public static HandRank empty(Hands hand) {
		return new HandRank(hand, List.empty());
	}
}

File: logic/src/main/kotlin/com/tp/holdem/logic/model/HandRank.kt
package com.tp.holdem.logic.model

import com.tp.holdem.logic.hands.HandRankComparator
import io.vavr.collection.List as VavrList

data class HandRank(
        internal val hand: Hands,
        internal val cardsThatMakeHand: VavrList<Card> = VavrList.empty()
) : Comparable<HandRank> {
    override fun compareTo(other: HandRank): Int {
        return HandRankComparator.compare(this, other)
    }

    override fun equals(other: Any?): Boolean {
        return other is HandRank && compareTo(other) == 0
    }

    companion object {
        @JvmStatic
        fun from(hand: Hands, cards: VavrList<Card>): HandRank {
            return HandRank(hand, cards)
        }

        @JvmStatic
        fun empty(hand: Hands): HandRank {
            return HandRank(hand, VavrList.empty())
        }
    }
}
